crates/ruff/src/rules/pylint/rules/redefined_loop_name.rs

```diff
@@ -0,0 +1,312 @@
+    let (outer_assignment_targets, inner_assignment_targets) = match node {
+        Node::Stmt(stmt) => match &stmt.node {
+            // With.
+            StmtKind::With { items, body, .. } => {
+                let outer_assignment_targets: Vec<ExprWithBindingKind<'a>> =
+                    assignment_targets_from_with_items(items, &checker.settings.dummy_variable_rgx)
+                        .map(|expr| ExprWithBindingKind {
+                            expr,
+                            binding_kind: BindingKind::With,
+                        })
+                        .collect();
+                let mut visitor = InnerForWithAssignTargetsVisitor {
+                    dummy_variable_rgx: &checker.settings.dummy_variable_rgx,
+                    assignment_targets: vec![],
+                };
+                for stmt in body {
+                    visitor.visit_stmt(stmt);
+                }
+                (outer_assignment_targets, visitor.assignment_targets)
+            }
+            // For and async for.
+            StmtKind::For { target, body, .. } | StmtKind::AsyncFor { target, body, .. } => {
+                let outer_assignment_targets: Vec<ExprWithBindingKind<'a>> =
+                    assignment_targets_from_expr(target, &checker.settings.dummy_variable_rgx)
+                        .map(|expr| ExprWithBindingKind {
+                            expr,
+                            binding_kind: BindingKind::For,
+                        })
+                        .collect();
+                let mut visitor = InnerForWithAssignTargetsVisitor {
+                    dummy_variable_rgx: &checker.settings.dummy_variable_rgx,
+                    assignment_targets: vec![],
+                };
+                for stmt in body {
+                    visitor.visit_stmt(stmt);
+                }
+                (outer_assignment_targets, visitor.assignment_targets)
+            }
+            _ => panic!(
+                "redefined_loop_name called on Statement that is not a With, For, or AsyncFor"
+            ),
+        },
+        Node::Expr(_) => panic!("redefined_loop_name called on Node that is not a Statement"),
+    };
+
+    for outer_assignment_target in &outer_assignment_targets {
+        let mut outer_name = None;
+        for inner_assignment_target in &inner_assignment_targets {
+            // Compare the targets structurally.
+            if ComparableExpr::from(outer_assignment_target.expr)
+                .eq(&(ComparableExpr::from(inner_assignment_target.expr)))
+            {
+                // Compute the textual name of the outer target lazily.
+                if outer_name.is_none() {
+                    outer_name = Some(
+                        checker
+                            .locator
+                            .slice(&Range::from_located(outer_assignment_target.expr)),
+                    );
+                }
```

Comment:
In that case, this could do slightly strange things:

```py
for a[
    "i"
] in y:
    pass
```

So in that case, we'd end up including the newlines in the violation message. But seems like an edge case...

Reply:
We _could_ instead use `unparse_expr` to turn the AST node into a string.

Reply:
I'm kind of partial to this -- what do you think?

```diff
diff --git a/crates/ruff/src/rules/pylint/rules/redefined_loop_name.rs b/crates/ruff/src/rules/pylint/rules/redefined_loop_name.rs
index ff538d30e..eb383358c 100644
--- a/crates/ruff/src/rules/pylint/rules/redefined_loop_name.rs
+++ b/crates/ruff/src/rules/pylint/rules/redefined_loop_name.rs
@@ -1,4 +1,5 @@
 use crate::ast::comparable::ComparableExpr;
+use crate::ast::helpers::unparse_expr;
 use crate::ast::types::{Node, Range};
 use crate::ast::visitor;
 use crate::ast::visitor::Visitor;
@@ -285,23 +286,14 @@ pub fn redefined_loop_name<'a, 'b>(checker: &'a mut Checker<'b>, node: &Node<'b>
     };

     for outer_assignment_target in &outer_assignment_targets {
-        let mut outer_name = None;
         for inner_assignment_target in &inner_assignment_targets {
             // Compare the targets structurally.
             if ComparableExpr::from(outer_assignment_target.expr)
                 .eq(&(ComparableExpr::from(inner_assignment_target.expr)))
             {
-                // Compute the textual name of the outer target lazily.
-                if outer_name.is_none() {
-                    outer_name = Some(
-                        checker
-                            .locator
-                            .slice(&Range::from_located(outer_assignment_target.expr)),
-                    );
-                }
                 checker.diagnostics.push(Diagnostic::new(
                     RedefinedLoopName {
-                        name: (*outer_name.unwrap()).to_string(),
+                        name: unparse_expr(outer_assignment_target.expr, checker.stylist),
                         outer_kind: outer_assignment_target.binding_kind,
                         inner_kind: inner_assignment_target.binding_kind,
                     },
```

Reply:
(I'm gonna commit this change in the interest of merging. If you feel strongly either way, let me know and I can always change in a follow-up.)

Reply:
Good point about the newlines - that works for me! (I didn't know about unparse_expr!) Now it gets computed once per violation. Oh, the inefficiency... 😆 Looks good, thanks!

Reply:
Haha I figured it was okay because we were already cloning once per violation?? >.<